kratos/utilities/strategies_utilities.h

```diff
@@ -0,0 +1,228 @@
+     * @param ScalingDiagonal The type of caling diagonal considered
+     * @return The scale norm
+     */
+    template<class TSparseSpace>
+    static double GetScaleNorm(
+        ModelPart& rModelPart,
+        typename TSparseSpace::MatrixType& rA,
+        const SCALING_DIAGONAL ScalingDiagonal = SCALING_DIAGONAL::NO_SCALING
+        )
+    {
+        switch (ScalingDiagonal) {
+            case SCALING_DIAGONAL::NO_SCALING:
+                return 1.0;
+            case SCALING_DIAGONAL::CONSIDER_PRESCRIBED_DIAGONAL: {
+                const ProcessInfo& r_current_process_info = rModelPart.GetProcessInfo();
+                KRATOS_ERROR_IF_NOT(r_current_process_info.Has(BUILD_SCALE_FACTOR)) << "Scale factor not defined at process info" << std::endl;
+                return r_current_process_info.GetValue(BUILD_SCALE_FACTOR);
+            }
+            case SCALING_DIAGONAL::CONSIDER_NORM_DIAGONAL:
+                return GetDiagonalNorm<TSparseSpace>(rA)/static_cast<double>(rA.size1());
+            case SCALING_DIAGONAL::CONSIDER_MAX_DIAGONAL:
+                return GetMaxDiagonal<TSparseSpace>(rA);
+//                 return TSparseSpace::TwoNorm(rA)/static_cast<double>(rA.size1());
+            default:
+                return GetMaxDiagonal<TSparseSpace>(rA);
+        }
+    }
+
+    /**
+     * @brief This method returns the diagonal norm considering for scaling the diagonal
+     * @param rA The LHS matrix
+     * @return The diagonal norm
+     */
+    template<class TSparseSpace>
+    static double GetDiagonalNorm(typename TSparseSpace::MatrixType& rA)
+    {
+        double* Avalues = rA.value_data().begin();
+        std::size_t* Arow_indices = rA.index1_data().begin();
+        std::size_t* Acol_indices = rA.index2_data().begin();
+
+        const double diagonal_norm = IndexPartition<std::size_t>(TSparseSpace::Size1(rA)).for_each<SumReduction<double>>([&](std::size_t Index){
+            const std::size_t col_begin = Arow_indices[Index];
+            const std::size_t col_end = Arow_indices[Index+1];
+            for (std::size_t j = col_begin; j < col_end; ++j) {
+                if (Acol_indices[j] == Index ) {
+                    return std::pow(Avalues[j], 2);
+                }
+            }
+            return 0.0;
+        });
+
+        return std::sqrt(diagonal_norm);
+    }
+
+    /**
+     * @brief This method returns the diagonal max value
+     * @param rA The LHS matrix
+     * @return The diagonal  max value
+     */
+    template<class TSparseSpace>
+    static double GetAveragevalueDiagonal(typename TSparseSpace::MatrixType& rA)
+    {
+        return 0.5 * (GetMaxDiagonal<TSparseSpace>(rA) + GetMinDiagonal<TSparseSpace>(rA));
+    }
+
+    /**
+     * @brief This method returns the diagonal max value
+     * @param rA The LHS matrix
+     * @return The diagonal  max value
+     */
+    template<class TSparseSpace>
+    static double GetMaxDiagonal(typename TSparseSpace::MatrixType& rA)
```

Comment:
```suggestion
    static double GetMaxDiagonal(const typename TSparseSpace::MatrixType& rA)
```

same for the other functions